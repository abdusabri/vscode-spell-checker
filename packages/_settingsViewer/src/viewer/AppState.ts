import {observable, computed} from 'mobx';
import { Settings, ConfigTarget, LocalId, SettingByConfigTarget, Config, Configs, LocalList, WorkspaceFolder, TextDocument, configTargets, ConfigTargets } from '../api/settings/';
import { normalizeCode, lookupCode } from '../iso639-1';
import { compareBy, compareEach } from '../api/utils/Comparable';
import { uniqueFilter } from '../api/utils';
import { Messenger } from '../api';


type Maybe<T> = T | undefined;
type TabTargets = ConfigTarget | 'file' | 'dictionaries' | 'about';

export interface Tab {
    label: string;
    target: TabTargets;
}

const tabs: Tab[] = [
    { label: 'User', target: 'user' },
    { label: 'Workspace', target: 'workspace' },
    { label: 'Folder', target: 'folder' },
    { label: 'File', target: 'file' },
    { label: 'Dictionaries', target: 'dictionaries' },
    { label: 'About', target: 'about' },
];

export interface LanguageInfo {
    code: string;
    name: string;
    dictionaries: string[];
    enabled: boolean;
}

export interface LanguageConfig {
    languages: LanguageInfo[];
    inherited?: ConfigTarget;
}

export interface LanguageConfigs extends SettingByConfigTarget<LanguageConfig> {}

export interface State {
    activeTabName: string;
    settings: Settings;
    tabs: Tab[];
    activeTab: Tab;
    languageConfig: LanguageConfigs;
}

export interface FoundInConfig<T> {
    value: Exclude<T, undefined>,
    target: ConfigTarget
}

type InheritedFromTarget<T> = {
    value: Exclude<T, undefined>,
    target: ConfigTarget;
}


type InheritMembers<T> = {
    [K in keyof T]: InheritedFromTarget<T[K]>;
}

type InheritedConfigs = SettingByConfigTarget<Config>;

export class AppState implements State {
    @observable activeTabName = '';
    @observable settings: Settings = {
        dictionaries: [
        ],
        configs: {
            user: { inherited: {}, locals: [], languageIdsEnabled: [] },
            workspace: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            folder: { inherited: { locals: 'user', languageIdsEnabled: 'user' }, locals: [], languageIdsEnabled: [] },
            file: undefined,
        }
    };
    @observable debugMode: boolean = false;

    constructor(private messageBus: Messenger) {}

    @computed get tabs() {
        const hidden = new Set<TabTargets>(configTargets.filter(target => !this.settings.configs[target]));
        if (this.workspaceFolders.length <= 1) {
            hidden.add(ConfigTargets.folder);
        }
        return tabs.filter(tab => !hidden.has(tab.target));
    }

    @computed get activeTab() {
        return this.tabs.find(t => t.label === this.activeTabName) || this.tabs[0];
    }

    @computed get languageConfig(): LanguageConfigs {
        const calcConfig = (target: ConfigTarget): LanguageConfig => {
            const config = this.settings.configs[target];
            const locals = config.locals; // todo: calc inheritance
            const inherited = config.inherited.locals;

            const infos = new Map<string, LanguageInfo>();

            const addLocalsToInfos = (locals: string[], dictionaryName: string | undefined) => {
                locals.map(normalizeCode).map(lookupCode).filter(notUndefined).forEach(lang => {
                    const { code, lang: language, country } = lang;
                    const name = country ? `${language} - ${country}` : language;
                    const found = this.isLocalEnabledEx(target, code);
                    const enabled = found && found.value || false;
                    const info: LanguageInfo = infos.get(name) || {
                        code,
                        name,
                        enabled,
                        dictionaries: [],
                    };
                    if (dictionaryName) {
                        info.dictionaries.push(dictionaryName);
                    }
                    infos.set(name, info);
                });
            }
            if (locals) {
                addLocalsToInfos(locals, undefined);
            }
            this.settings.dictionaries.forEach(dict => addLocalsToInfos(dict.locals, dict.name));

            return {
                languages: [...infos.values()].sort(compareEach(
                    compareBy(info => !info.dictionaries.length),
                    compareBy('name'),
                )),
                inherited
            };
        };

        return {
            user: calcConfig('user'),
            workspace: calcConfig('workspace'),
            folder: calcConfig('folder'),
        }
    }

    @computed get activeTabIndex(): number {
        const index = this.tabs.findIndex(t => t.label === this.activeTabName);
        return index > 0 ? index : 0;
    }

    @computed get workspaceFolderNames(): string[] {
        const workspace = this.settings.workspace;
        const folders = workspace && workspace.workspaceFolders || [];
        return folders.map(f => f.name);
    }

    @computed get activeWorkspaceFolder(): string | undefined {
        const folder = this.findMatchingFolder(this.settings.activeFolderUri);
        return folder && folder.name;
    }

    @computed get workspaceFolders(): WorkspaceFolder[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.workspaceFolders || [];
    }

    @computed get activeFileUri(): string | undefined {
        return this.settings.activeFileUri;
    }

    @computed get documents(): TextDocument[] {
        const workspace = this.settings.workspace;
        return workspace && workspace.textDocuments || [];
    }

    @computed get documentSelection(): { label: string, value: string }[] {
        return this.documents.map(doc => ({ label: doc.fileName, value: doc.uri }));
    }

    private findMatchingFolder(uri: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.uri === uri)[0];
    }

    private findMatchingFolderByName(name: string | undefined): WorkspaceFolder | undefined {
        return this.workspaceFolders.filter(f => f.name === name)[0];
    }

    actionSetLocal(target: ConfigTarget, local: LocalId, enable: boolean) {
        const localsCurrent = this.settings.configs[target].locals;
        const locals = (enable ? [...localsCurrent, local] : localsCurrent.filter(v => v !== local)).filter(uniqueFilter());
        this.settings.configs[target].locals = locals;
        this.messageBus.postMessage({ command: 'EnableLocalMessage', value: { target, local, enable } });
    }

    actionSetDebugMode(isEnabled: boolean) {
        this.debugMode = isEnabled;
    }

    private isLocalEnabledEx(field: ConfigTarget, code: LocalId):InheritedFromTarget<boolean> {
        const config = this.settings.configs[field];
        const target = config.inherited.locals || field;
        const locals = config.locals;
        return  {
            value: locals.map(normalizeCode).includes(code),
            target
        };
    }


    actionActivateTabIndex(index: number) {
        const tab = this.tabs[index];
        if (tab) {
            this.actionActivateTab(tab.label);
        }
    }

    actionActivateTab(tabName: string) {
        this.activeTabName = tabName;
        this.messageBus.postMessage({ command: 'SelectTabMessage', value: this.activeTabName });
    }

    actionSelectFolder(folderName: string) {
        const folder = this.findMatchingFolderByName(folderName);
        const folderUri = folder && folder.uri;
        this.settings.activeFolderUri = folderUri;
        if (folderUri !== undefined) {
            this.messageBus.postMessage({ command: 'SelectFolderMessage', value: folderUri })
        }
    }

    actionSelectDocument(documentUri: string) {
        this.messageBus.postMessage({ command: 'SelectFileMessage', value: documentUri });
    }

    actionEnableLanguageId(languageId: string, enable: boolean, target?: ConfigTarget) {
        const fileConfig = this.settings.configs.file;
        if (fileConfig && fileConfig.languageId === languageId) {
            fileConfig.languageEnabled = enable;
        }
        this.messageBus.postMessage({ command: 'EnableLanguageIdMessage', value: { languageId, enable, target }});
    }
}

function notUndefined<T>(a : T): a is Exclude<T, undefined> {
    return a !== undefined;
}


