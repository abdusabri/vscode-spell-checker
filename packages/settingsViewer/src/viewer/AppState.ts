import {observable, computed} from 'mobx';
import { Settings } from '../api/settings/';

export const cats = [
    { title: 'Coding Cat', image: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif', icon: 'home'},
    { title: 'Compiling Cat', image: 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif', icon: 'face'},
    { title: 'Testing Cat', image: 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif', icon: 'favorite'},
];

export const tabLabels = ['Language'].concat(cats.map(cat => cat.title));

export class AppState {
    @observable activeTabIndex = 0;
    @observable timer = 0;
    @observable counter = 0;
    @observable settings: Settings = {
        locals: [
            {
                code: 'en',
                name: 'English',
                dictionaries: ['en', 'en-us'],
                enabled: true,
                isInUserSettings: true,
            },
            {
                code: 'es',
                name: 'Spanish',
                dictionaries: ['es', 'es-ES'],
                enabled: true,
                isInUserSettings: true,
            },
        ],
    };

    @computed get activeTab() {
        return tabLabels[this.activeTabIndex];
    }

    constructor() {
        setInterval(() => {
            this.timer += 1;
        }, 1000);
    }

    resetTimer() {
        this.timer = 0;
    }
}
