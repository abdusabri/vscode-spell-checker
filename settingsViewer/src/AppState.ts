import {observable, computed} from 'mobx';
import { ConfigurationForDocument } from './settings/';

export const cats = [
    { title: 'Coding Cat', image: 'https://media.giphy.com/media/JIX9t2j0ZTN9S/giphy.gif', icon: 'home'},
    { title: 'Compiling Cat', image: 'https://media.giphy.com/media/mlvseq9yvZhba/giphy.gif', icon: 'face'},
    { title: 'Testing Cat', image: 'https://media.giphy.com/media/3oriO0OEd9QIDdllqo/giphy.gif', icon: 'favorite'},
];

export const tabLabels = ['Language'].concat(cats.map(cat => cat.title));

export class AppState {
    @observable activeTabIndex = 0;
    @observable timer = 0;
    @observable counter = 0;
    @observable config: ConfigurationForDocument = {
        languageEnabled: true,
        fileEnabled: true,
        settings: {},
        docSettings: {}
    };
    @computed get activeTab() {
        return tabLabels[this.activeTabIndex];
    }

    constructor() {
        setInterval(() => {
            this.timer += 1;
        }, 1000);
    }

    resetTimer() {
        this.timer = 0;
    }
}
