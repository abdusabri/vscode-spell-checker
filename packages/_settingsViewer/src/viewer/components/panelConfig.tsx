import * as React from 'react';
import {observer} from 'mobx-react';
import { AppState } from '../AppState';
import { ConfigTarget, ConfigTargets } from '../../api/settings';
import { SectionLanguage } from './sectionLanguage';
import { SectionFiletypes } from './sectionFiletypes';
import Select from '@material/react-select';

@observer
export class PanelConfig extends React.Component<{appState: AppState, target: ConfigTarget}, {}> {
    render() {
        const appState = this.props.appState;
        const target = this.props.target;
        const settings = appState.settings;
        const config = settings.configs[target];
        if (!config) {
            return <div></div>
        }
        const workspace = settings.workspace;
        const workspaceFolders = workspace && workspace.workspaceFolders || [];
        const options = workspaceFolders.map(folder => folder.name);
        return (
            <div>
                {target === ConfigTargets.folder
                ? <Select className='select_folder'
                        label='Folder'
                        options={options}
                        onChange={(evt) => this.props.appState}
                        value={options[0]}
                    />
                : ''}
                <SectionLanguage appState={appState} target={target}></SectionLanguage>
                <SectionFiletypes appState={appState} target={target}></SectionFiletypes>
            </div>
        );
     }
}
