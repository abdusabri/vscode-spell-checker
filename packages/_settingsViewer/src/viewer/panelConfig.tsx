import * as React from 'react';
import {Cell, Grid, Row} from '@material/react-layout-grid';
import { AppState } from './AppState';
import * as spellCheckIcon from './images/SpellCheck.xs.png';
import {tf} from '../api/utils';
import { LocalInfo } from './AppState';
import { LocalSetting, LocalId, ConfigTarget } from '../api/settings';

type LocalSettingField = keyof LocalSetting;


export class PanelConfig extends React.Component<{appState: AppState, target: ConfigTarget}, {}> {
    render() {
        const appState = this.props.appState;
        const target = this.props.target;
        const config = appState.settings.configs[target];
        if (!config) {
            return <div></div>
        }
        return (
            <Grid>
                <Row>
                    <Cell columns={12}>
                        <h3>Language</h3>
                        <div>
                        {config.locals ? config.locals.join(', ') : '- none -'}
                        </div>
                    </Cell>
                </Row>
                <Row>
                    <Cell columns={12}>
                        <h3>File Types</h3>
                        <div>
                        {config.fileTypesEnabled ? config.fileTypesEnabled.join(', ') : '- none -'}
                        </div>
                    </Cell>
                </Row>
            </Grid>
        );
     }
}
