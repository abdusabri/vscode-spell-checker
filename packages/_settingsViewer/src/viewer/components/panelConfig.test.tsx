
import * as React from 'react';
import { AppState } from '../AppState';
import { create } from 'react-test-renderer';
import { sampleSettings } from '../samples/sampleSettings';
import { PanelConfig } from './panelConfig';

describe('Config Panel Verification', () => {
    it('tests the snapshot', () => {
        const appState = getSampleAppState();
        const panelRenderer = create(<PanelConfig appState={appState} target='user'></PanelConfig>).toJSON()!;
        expect(panelRenderer).toMatchSnapshot();
    });

    function getSampleAppState(): AppState {
        const appState: AppState = new AppState();
        appState.settings = sampleSettings;
        return appState;
    }
});
