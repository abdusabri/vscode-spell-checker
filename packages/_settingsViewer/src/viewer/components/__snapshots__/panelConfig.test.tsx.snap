// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Panel Verification tests the snapshot 1`] = `
<div>
  <div
    className="mdc-select select_folder"
  >
    <select
      className="mdc-select__native-control"
      disabled={false}
      label="Folder"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      options={
        Array [
          "ripgrep",
          "dutch",
        ]
      }
      value="ripgrep"
    >
      <option
        value="ripgrep"
      >
        ripgrep
      </option>
      <option
        value="dutch"
      >
        dutch
      </option>
    </select>
    <label
      className="mdc-floating-label mdc-floating-label--float-above"
      onAnimationEnd={[Function]}
    >
      Folder
    </label>
    <div
      className="mdc-line-ripple"
      onTransitionEnd={[Function]}
      style={Object {}}
    />
  </div>
  <div>
    <h2>
      Language 
      
    </h2>
    <div>
      <ul
        aria-orientation="vertical"
        className="mdc-list mdc-list--two-line"
      >
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="checkbox"
          tabIndex={0}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              Czech
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              cs-cz
            </span>
          </span>
          <div
            className="mdc-checkbox mdc-checkbox--upgraded mdc-list-item__meta"
            onAnimationEnd={[Function]}
            tabIndex={-1}
          >
            <input
              aria-checked={false}
              checked={false}
              className="mdc-checkbox__native-control"
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            <div
              className="mdc-checkbox__background"
            >
              <svg
                className="mdc-checkbox__checkmark"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  className="mdc-checkbox__checkmark-path"
                  d="M1.73,12.91 8.1,19.28 22.79,4.59"
                  fill="none"
                />
              </svg>
              <div
                className="mdc-checkbox__mixedmark"
              />
            </div>
          </div>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="checkbox"
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              English
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              en_us
            </span>
          </span>
          <div
            className="mdc-checkbox mdc-checkbox--upgraded mdc-list-item__meta"
            onAnimationEnd={[Function]}
            tabIndex={-1}
          >
            <input
              aria-checked={false}
              checked={true}
              className="mdc-checkbox__native-control"
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            <div
              className="mdc-checkbox__background"
            >
              <svg
                className="mdc-checkbox__checkmark"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  className="mdc-checkbox__checkmark-path"
                  d="M1.73,12.91 8.1,19.28 22.79,4.59"
                  fill="none"
                />
              </svg>
              <div
                className="mdc-checkbox__mixedmark"
              />
            </div>
          </div>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="checkbox"
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              English - United Kingdom
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              en-gb
            </span>
          </span>
          <div
            className="mdc-checkbox mdc-checkbox--upgraded mdc-list-item__meta"
            onAnimationEnd={[Function]}
            tabIndex={-1}
          >
            <input
              aria-checked={false}
              checked={false}
              className="mdc-checkbox__native-control"
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            <div
              className="mdc-checkbox__background"
            >
              <svg
                className="mdc-checkbox__checkmark"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  className="mdc-checkbox__checkmark-path"
                  d="M1.73,12.91 8.1,19.28 22.79,4.59"
                  fill="none"
                />
              </svg>
              <div
                className="mdc-checkbox__mixedmark"
              />
            </div>
          </div>
        </li>
        <li
          className="mdc-list-item"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="checkbox"
          tabIndex={-1}
        >
          <i
            className="material-icons mdc-list-item__graphic"
            tabIndex={-1}
          >
            import_contacts
          </i>
          <span
            className="mdc-list-item__text"
            tabIndex={-1}
          >
            <span
              className="mdc-list-item__primary-text"
              tabIndex={-1}
            >
              English - United States
            </span>
            <span
              className="mdc-list-item__secondary-text"
              tabIndex={-1}
            >
              en_us
            </span>
          </span>
          <div
            className="mdc-checkbox mdc-checkbox--upgraded mdc-list-item__meta"
            onAnimationEnd={[Function]}
            tabIndex={-1}
          >
            <input
              aria-checked={false}
              checked={false}
              className="mdc-checkbox__native-control"
              disabled={false}
              onChange={[Function]}
              type="checkbox"
            />
            <div
              className="mdc-checkbox__background"
            >
              <svg
                className="mdc-checkbox__checkmark"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  className="mdc-checkbox__checkmark-path"
                  d="M1.73,12.91 8.1,19.28 22.79,4.59"
                  fill="none"
                />
              </svg>
              <div
                className="mdc-checkbox__mixedmark"
              />
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
  <div>
    <h2>
      File Types
    </h2>
    <div>
      asciidoc, c, cfml, cpp, csharp, css, go, handlebars, html, jade, java, javascript, javascriptreact, json, latex, less, markdown, php, plaintext, pub, python, restructuredtext, rust, scala, scss, text, typescript, typescriptreact, yaml, yml
    </div>
  </div>
</div>
`;
